src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs

```diff
@@ -0,0 +1,224 @@
+                if (importCompletionOptionValue == false ||
+                    (importCompletionOptionValue == null && !IsExperimentEnabled(document.Project.Solution.Workspace)))
+                {
+                    return;
+                }
+            }
+
+            // Find all namespaces in scope at current cursor location, 
+            // which will be used to filter so the provider only returns out-of-scope types.
+            var namespacesInScope = GetNamespacesInScope(document, syntaxContext, cancellationToken);
+            await AddCompletionItemsAsync(completionContext, syntaxContext, namespacesInScope, isExpandedCompletion, cancellationToken).ConfigureAwait(false);
+        }
+
+        private HashSet<string> GetNamespacesInScope(Document document, SyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            var semanticModel = syntaxContext.SemanticModel;
+            var importedNamespaces = GetImportedNamespaces(syntaxContext.LeftToken.Parent, semanticModel, cancellationToken);
+
+            // This hashset will be used to match namespace names, so it must have the same case-sensitivity as the source language.
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>()!;
+            var namespacesInScope = new HashSet<string>(importedNamespaces, syntaxFacts.StringComparer);
+
+            // Get containing namespaces.
+            var namespaceSymbol = semanticModel.GetEnclosingNamespace(syntaxContext.Position, cancellationToken);
+            while (namespaceSymbol != null)
+            {
+                namespacesInScope.Add(namespaceSymbol.ToDisplayString(SymbolDisplayFormats.NameFormat));
+                namespaceSymbol = namespaceSymbol.ContainingNamespace;
+            }
+
+            return namespacesInScope;
+        }
+
+        internal override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem completionItem, TextSpan completionListSpan, char? commitKey, CancellationToken cancellationToken)
+        {
+            var containingNamespace = ImportCompletionItem.GetContainingNamespace(completionItem);
+
+            if (await ShouldCompleteWithFullyQualifyTypeName().ConfigureAwait(false))
+            {
+                var fullyQualifiedName = $"{containingNamespace}.{completionItem.DisplayText}";
+                var change = new TextChange(completionListSpan, fullyQualifiedName);
+
+                return CompletionChange.Create(change);
+            }
+
+            // Find context node so we can use it to decide where to insert using/imports.
+            var tree = (await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false))!;
+            var root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var addImportContextNode = root.FindToken(completionListSpan.Start, findInsideTrivia: true).Parent;
```

Comment:
do we have tests inside trivia?

Reply:
I thought we do, but turns out we don't

Reply:
plz add :)

Reply:
Turns out we will not be inside trivia since `cref` is not a "type context", I'm fine with this for now. A test is added.